### PR TITLE
Chore: Change *RelabelConfigs to values instead of Pointers 

### PR DESCRIPTION
**Pull request by mviswanathsai**
## Description
Relabel configs previously were of type `[]*monitoringv1.RelabelConfigs`, this PR changes that to `[]monitoringv1.RelabelConfigs`. 
Earlier users could write invalid manifests like this one:
```
relabelConfigs:
- null
- sourceLabels: ...
```

Which the operator would then have to validate and then reject. By changing the type from pointer to just the value, we are able to reject such invalid manifests directly without having to reject it at the operator level.

This PR is to check the viability of this change. 
Related discussion: https://kubernetes.slack.com/archives/C01B03QCSMN/p1712306476986909, [#6379 (comment)](https://github.com/prometheus-operator/prometheus-operator/pull/6379#discussion_r1521132751)

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [ ] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [x] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Verification

Please check the [Prometheus-Operator testing guidelines](../TESTING.md) for recommendations about automated tests.

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```
Change type of the different RelabelConfigs fields to `[]monitoringv1.RelabelConfigs` instead of an array of pointers
```

### CODE REVIEW

**Comment by mviswanathsai** [comment]
@simonpasquier the tests do not seem to be failing. Nice!

File: pkg/prometheus/resource_selector.go [hunk 1]
```diff
@@ -247,13 +247,14 @@ func validateScrapeIntervalAndTimeout(p monitoringv1.PrometheusInterface, scrape
 	return CompareScrapeTimeoutToScrapeInterval(scrapeTimeout, scrapeInterval)
 }
 
-func validateRelabelConfigs(p monitoringv1.PrometheusInterface, rcs []*monitoringv1.RelabelConfig) error {
+func validateRelabelConfigs(p monitoringv1.PrometheusInterface, rcs []monitoringv1.RelabelConfig) error {
 	for i, rc := range rcs {
-		if rc == nil {
-			return fmt.Errorf("null relabel config")
-		}
+		// No need for this now
```
Comment (nicolastakashi) [inline pkg/prometheus/resource_selector.go:252]: (nit) remove the commented coded pls 😄

Comment (mviswanathsai) [reply pkg/prometheus/resource_selector.go]: Done!

**Review by ArthurSens — COMMENTED** [review]
I'm a bit concerned about performance issues 😬 

Using raw values or pointer values as function arguments, e.g. when validating relabel configs or generating the actual prometheus configuration, will have different behaviors in terms of memory allocations.

The go compiler can make several optimizations for simple functions([Function inlining](https://go.dev/wiki/CompilerOptimizations#escape-analysis-and-inlining)), avoiding the need to copy arguments in new memory registries but it won't do it for complex functions.

If inlining is not happening, during runtime we will copy function arguments into new memory registries when they are called. A pointer is just 8 bytes, while complex raw structs will take as much size as the struct requires.

---

TLDR; I think it would be a good idea to benchmark functions that uses RelabelConfigs as arguments before committing to this change. It's not a bad idea to maintain a single nilcheck if that means less memory allocations

**Comment by ArthurSens** [comment]
I just did the benchmark myself:

<details><summary>Benchmark code</summary>

```go
func BenchmarkSelectServiceMonitors(b *testing.B) {
	ca, err := os.ReadFile("../../test/e2e/remote_write_certs/ca.crt")
	require.NoError(b, err)

	cert, err := os.ReadFile("../../test/e2e/remote_write_certs/client.crt")
	require.NoError(b, err)

	key, err := os.ReadFile("../../test/e2e/remote_write_certs/client.key")
	require.NoError(b, err)

	cs := fake.NewSimpleClientset(
		&v1.Secret{
			ObjectMeta: metav1.ObjectMeta{
				Name:      "secret",
				Namespace: "test",
			},
			Data: map[string][]byte{
				"ca":         ca,
				"invalid_ca": []byte("garbage"),
				"cert":       cert,
				"key":        key,
			},
		},
		&v1.ConfigMap{
			ObjectMeta: metav1.ObjectMeta{
				Name:      "configmap",
				Namespace: "test",
			},
			Data: map[string]string{
				"ca":   string(ca),
				"cert": string(cert),
			},
		},
	)

	rs := NewResourceSelector(
		newLogger(),
		&monitoringv1.Prometheus{
			Spec: monitoringv1.PrometheusSpec{
				CommonPrometheusFields: monitoringv1.CommonPrometheusFields{},
			},
		},
		assets.NewStore(cs.CoreV1(), cs.CoreV1()),
		nil,
		operator.NewMetrics(prometheus.NewPedanticRegistry()),
		record.NewFakeRecorder(1),
	)

	// Some common K8s relabeling configs
	rcs := make([]*monitoringv1.RelabelConfig, 15)
	for i := range rcs {
		rcs[i] = &monitoringv1.RelabelConfig{
			SourceLabels: []monitoringv1.LabelName{"__name__"},
			Action:       "drop",
			Regex:        "kubelet_(pod_worker_latency_microseconds|pod_start_latency_microseconds|cgroup_manager_latency_microseconds|pod_worker_start_latency_microseconds|pleg_relist_latency_microseconds|pleg_relist_interval_microseconds|runtime_operations|runtime_operations_latency_microseconds|runtime_operations_errors|eviction_stats_age_microseconds|device_plugin_registration_count|device_plugin_alloc_latency_microseconds|network_plugin_operations_latency_microseconds)",
		}
	}

	sm := &monitoringv1.ServiceMonitor{
		ObjectMeta: metav1.ObjectMeta{
			Name:      "test",
			Namespace: "test",
		},
		Spec: monitoringv1.ServiceMonitorSpec{
			Endpoints: []monitoringv1.Endpoint{
				{
					RelabelConfigs: rcs,
				},
			},
		},
	}

	b.ResetTimer()

	for i := 0; i < b.N; i++ {
		_, err := rs.SelectServiceMonitors(context.Background(), func(_ string, _ labels.Selector, appendFn cache.AppendFunc) error {
			appendFn(sm)
			return nil
		})
		require.NoError(b, err)
	}
}

```

</details>

It appears what I said above is incorrect or irrelevant when compared to everything else used in SelecServiceMonitors :)

Results:
```console
goos: darwin
goarch: arm64
pkg: github.com/prometheus-operator/prometheus-operator/pkg/prometheus
                         │ ../prometheus-operator/bench-with-pointers.txt │ ../prometheus-operator/bench-without-pointers.txt │
                         │                     sec/op                     │             sec/op               vs base          │
SelectServiceMonitors-12                                      653.3µ ± 4%                       655.4µ ± 2%  ~ (p=0.684 n=10)

                         │ ../prometheus-operator/bench-with-pointers.txt │ ../prometheus-operator/bench-without-pointers.txt │
                         │                      B/op                      │            B/op             vs base               │
SelectServiceMonitors-12                                     1.638Mi ± 0%                 1.638Mi ± 0%  -0.00% (p=0.000 n=10)

                         │ ../prometheus-operator/bench-with-pointers.txt │ ../prometheus-operator/bench-without-pointers.txt │
                         │                   allocs/op                    │         allocs/op           vs base               │
SelectServiceMonitors-12                                      4.408k ± 0%                  4.393k ± 0%  -0.34% (p=0.000 n=10)

```

**Review by ArthurSens — DISMISSED** [review]
LGTM

**Comment by ArthurSens** [comment]
Sorry @mviswanathsai, I had some problems in the last week and couldn't keep up with all the activity in the project.

Could you rebase the PR when you have the time? It should be good to merge afterwards

**Review by ArthurSens — APPROVED** [review]
